Cython/Compiler/ExprNodes.py

```diff
@@ -12414,7 +12419,7 @@ def infer_builtin_types_operation(self, type1, type2):
 
     def analyse_operation(self, env):
         self._check_truedivision(env)
-        NumBinopNode.analyse_operation(self, env)
+        result = NumBinopNode.analyse_operation(self, env)
         if self.is_cpp_operation():
             self.cdivision = True
         if not self.type.is_pyobject:
```

Comment:
This code only (safely) works if `NumBinopNode.analyse_operation()` returns exactly `self`, not anything else. That is not an assumption we should generally make. It's really the whole point of returning a node that it allows us to return a *different* node.

It's difficult to deal with this correctly, though. It's ambiguous if the base method returns a different node or even different type of node. It might be a wrapper around the original (self) node, or it might have replaced the node entirely. We can't know this here. Not sure how to handle this here.

Reply:
The real assumption here is that `result` contains `self` (i.e. `self` is still relevant and can be manipulated, even if it's wrapped in a coercion node or something like that).

I've added a comment and an assert to clarify/enforce that.

I'm not hugely happy with this either though, but I think it's better than it initially looks.